main/http_server/http_server.c

```diff
@@ -61,6 +61,100 @@ typedef struct rest_server_context
 
 #define CHECK_FILE_EXTENSION(filename, ext) (strcasecmp(&filename[strlen(filename) - strlen(ext)], ext) == 0)
 
+static esp_err_t ip_in_private_range(uint32_t ip){
+        //Private IP ranges (little endian, 192.168.0.0 => 0.0.168.192)
+    //192.168.0.0
+    uint32_t sixteen_bit_block = 0b00000000000000001010100011000000;
+    uint32_t sixteen_bit_mask =  0b00000000000000001111111111111111;
+
+    if((ip & sixteen_bit_mask) == sixteen_bit_block){
+         return ESP_OK;
+    }
+    //172.16.0.0 
+    uint32_t twenty_bit_block = 0b00000000000000000001000010101100;
+    uint32_t twenty_bit_mask = 0b00000000000000001111000011111111;
+    if((ip & twenty_bit_mask) == twenty_bit_block){
+         return ESP_OK;
+    }
+    //10.0.0.0
+    uint32_t twenty_four_bit_block = 0b00000000000000000000000000001010;
+    uint32_t twenty_four_bit_mask = 0b00000000000000000000000011111111;
+    if((ip & twenty_four_bit_mask) == twenty_four_bit_block){
+         return ESP_OK;
+    }
+
+    return ESP_FAIL;
+}
+static esp_err_t check_is_same_network(httpd_req_t * req){
+
+    int sockfd = httpd_req_to_sockfd(req);
+    char ipstr[INET6_ADDRSTRLEN];
+    struct sockaddr_in6 addr;   // esp_http_server uses IPv6 addressing
+    socklen_t addr_size = sizeof(addr);
+    
+    if (getpeername(sockfd, (struct sockaddr *)&addr, &addr_size) < 0) {
+        ESP_LOGE(TAG, "Error getting client IP");
+        return ESP_FAIL;
+    }
+
+
+    uint32_t request_ip_addr = addr.sin6_addr.un.u32_addr[3];
+
+    // // Convert to IPv6 string
+    // inet_ntop(AF_INET, &addr.sin6_addr, ipstr, sizeof(ipstr));
+
+    // Convert to IPv4 string
+    inet_ntop(AF_INET, &request_ip_addr, ipstr, sizeof(ipstr));
+
+   
+
+    char origin[128]; 
+    char ip_str[16];  // Buffer to hold the extracted IP address string
+    uint32_t origin_ip_addr = 0;
+    // Attempt to get the Origin header
+    if (httpd_req_get_hdr_value_str(req, "Origin", origin, sizeof(origin)) == ESP_OK) {
+        ESP_LOGI("CORS", "Origin header: %s", origin);
+
+        // Find the start of the IP address in the Origin header
+        const char *prefix = "http://";
+        char *ip_start = strstr(origin, prefix);
+        if (ip_start) {
+            ip_start += strlen(prefix); // Move past "http://"
+
+            // Extract the IP address portion (up to the next '/')
+            char *ip_end = strchr(ip_start, '/');
+            size_t ip_len = ip_end ? (size_t)(ip_end - ip_start) : strlen(ip_start);
+            if (ip_len < sizeof(ip_str)) {
+                strncpy(ip_str, ip_start, ip_len);
+                ip_str[ip_len] = '\0'; // Null-terminate the string
+
+                // Convert the IP address string to uint32_t
+                origin_ip_addr = inet_addr(ip_str);
+                if (origin_ip_addr == INADDR_NONE) {
+                    ESP_LOGW("CORS", "Invalid IP address: %s", ip_str);
+                } else {
+                    ESP_LOGI("CORS", "Extracted IP address %lu", origin_ip_addr);
```

Comment:
We could probably change the happy path to DEBUG log lines. There is a small performance impact by all the additional logging.
> ESP_LOGI("CORS", "Origin header: %s", origin);
> ESP_LOGI("CORS", "Extracted IP address %lu", origin_ip_addr);